### PR TITLE
First stab at adding footnote text next to the ref

### PR DESCRIPTION
**Pull request by DmitrySharabin**

### CODE REVIEW

File: src/markdown.js [hunk 1]
```diff
@@ -43,6 +43,42 @@ export default function (config, options = {}) {
 	if (options.footnotes !== false) {
 		md.use(markdownItFootnote, options.footnotes);
 
+		let allFootnotes = new Map(); // pageId → { label: footnote }
+
+		function extractFootnoteTexts (tokens, pageId) {
+			let footnotes = {};
```
Comment (LeaVerou) [inline src/markdown.js]: Won't this erase previous footnotes if `extractFootnoteTexts is called multiple times on the same page?

File: src/markdown.js [hunk 2]
```diff
@@ -67,17 +102,24 @@ export default function (config, options = {}) {
...
+
+			let footnotes = allFootnotes.get(env.id);
+			let footnoteLabel = tokens[idx].meta.label;
+			let footnoteText = footnotes[footnoteLabel];
+			footnoteText = md.renderInline(footnoteText);
+			footnoteText = `<span class="footnote-text">${footnoteText}</span>`
 
-		// 	return `<sup class="footnote-ref"><a href="#fn${id}" id="fnref${refid}">${caption}</a></sup>`
-		// };
+			return `<sup class="footnote-ref"><a epub:type="noteref" href="#fn${id}" id="fnref${refId}">${caption}</a>${footnoteText}</sup>`
```
Comment (LeaVerou) [inline src/markdown.js]:
```suggestion
			return `<sup class="footnote-ref"><a epub:type="noteref" href="#fn${id}" id="fnref${refId}">${caption}</a></sup>${footnoteText}`
```

File: src/markdown.js [hunk 3]
```diff
@@ -67,17 +102,24 @@ export default function (config, options = {}) {
...
+			let refId = id;
+			if (tokens[idx].meta.subId > 0) {
+				refId += `:${tokens[idx].meta.subId}`;
+			}
+
+			let footnotes = allFootnotes.get(env.id);
+			let footnoteLabel = tokens[idx].meta.label;
+			let footnoteText = footnotes[footnoteLabel];
+			footnoteText = md.renderInline(footnoteText);
+			footnoteText = `<span class="footnote-text">${footnoteText}</span>`
```
Comment (LeaVerou) [inline src/markdown.js]:
```suggestion
			footnoteText = `<span class="footnote-text inline" hidden>${footnoteText}</span>`
```

(We’ll make it visible with CSS when needed)

File: src/markdown.js [hunk 4]
```diff
@@ -67,17 +102,24 @@ export default function (config, options = {}) {
...
-		// 	if (tokens[idx].meta.subId > 0) {
-		// 		refid += `:${tokens[idx].meta.subId}`
-		// 	}
+		md.renderer.rules.footnote_ref = function (tokens, idx, options, env, slf) {
+			let id = slf.rules.footnote_anchor_name(tokens, idx, options, env, slf);
+			let caption = slf.rules.footnote_caption(tokens, idx, options, env, slf);
+
+			let refId = id;
+			if (tokens[idx].meta.subId > 0) {
+				refId += `:${tokens[idx].meta.subId}`;
```
Comment (LeaVerou) [inline src/markdown.js]: What does that do and why? (add comments, don't reply here)

File: src/markdown.js [hunk 5]
```diff
@@ -43,6 +43,42 @@ export default function (config, options = {}) {
 	if (options.footnotes !== false) {
 		md.use(markdownItFootnote, options.footnotes);
 
+		let allFootnotes = new Map(); // pageId → { label: footnote }
+
+		function extractFootnoteTexts (tokens, pageId) {
```
Comment (LeaVerou) [inline src/markdown.js:55]: Add comments on what this does and why we need to do this

File: src/markdown.js (hunk 4 above)
Comment (DmitrySharabin) [reply src/markdown.js]: Actually, I don't know. This code is a copy of the footnotes plugin with the necessary changes. So... I don't know what to write in the comment 🙈

Comment (LeaVerou) [reply src/markdown.js]: You write exactly that. `// Code from <link to code on github>`. But do we really need to copy? Maybe we can store a reference to the existing function and monkey patch? Once I have that link I can tell you

File: src/markdown.js (hunk 1 above)
Comment (LeaVerou) [inline src/markdown.js]:
```suggestion
			let footnotes = allFootnotes.get(pageId) ?? {};
```

File: src/markdown.js (hunk 4 above)
Comment (DmitrySharabin) [reply src/markdown.js]: Fair enough. Done!

Comment (LeaVerou) [reply src/markdown.js]: Looks like we can wrap the function and just append to its return value?

Comment (DmitrySharabin) [reply src/markdown.js]: Oh, good idea! I need to try this out

**Comment by DmitrySharabin** [comment]
Thank you so much for your feedback. I tried to address all of it. Could you please give the PR another shot?

File: src/markdown.js [hunk 6]
```diff
@@ -43,6 +43,50 @@ export default function (config, options = {}) {
...
+					done = false;
+					label = token.meta.label;
+					footnotes[label] = "";
+				}
+				else if (token.type === "footnote_close") {
+					done = true;
+				}
+				else if (done === false && token.content) {
+					// Tokens of all other types after “footnote_open” till “footnote_close” form a footnote text
+					footnotes[label] += token.content;
```
Comment (LeaVerou) [inline src/markdown.js:71]: Have you tested this with footnote text that contains other tokens?

File: src/markdown.js [hunk 7]
```diff
@@ -43,6 +43,50 @@ export default function (config, options = {}) {
...
+		// Override the render method to capture footnote texts
+		md.render = function (src, env) {
+			let tokens = md.parse(src, env);
+
+			let pageId = env.id;
+			if (pageId) {
+				extractFootnoteTexts(tokens, pageId);
+			}
+
+			return md.renderer.render(tokens, md.options, env);
```
Comment (LeaVerou) [inline src/markdown.js]: Wait, I just noticed this. Isn't this the general MD renderer used for everything? Can't we override something more specific? What happens if two plugins do this?

File: src/markdown.js (hunk 6 above)
Comment (DmitrySharabin) [reply src/markdown.js:71]: If you mean the case when a footnote text has markdown markup (parsed as separate tokens), then yes, I tested it. That's why we render the footnote text before adding it to the footnote ref:

image

Did I get your question right?

File: src/markdown.js (hunk 7 above)
Comment (DmitrySharabin) [reply src/markdown.js]: On my first try, I monkey-patched the original `md.render` method by performing the needed operations and returning `md.render.apply(md, args)`.

The issue I faced pushed me to reject this approach: When we do `md.parse()` to get the tokens, we tell MD about footnotes, so when the footnotes plugin actually processes them, we have two copies of the footnotes.

That's why I ended up with this approach. I can give it another look. Though, if you have any ideas on how to fix that, I'm open to any suggestions.

Comment (DmitrySharabin) [reply src/markdown.js]: What if we get another (clone/copy) MD instance (inside `md.render()`) to get the tokens so the original one wouldn't know about them? Is it too much?

Comment (DmitrySharabin) [reply src/markdown.js]: Okay, I went with another approach. :)

File: src/markdown.js (hunk 6 above)
Comment (LeaVerou) [reply src/markdown.js:71]: Yes! It's unfortunate we're rendering the same piece of content multiple times, but what can we do.

**Review by LeaVerou — APPROVED** [review]

**Pull request merged**